src/games/rcll/net.clp

```diff
@@ -274,12 +275,80 @@
 	)
 )
 
+(defrule net-recv-Workpiece
+  ?mf <- (protobuf-msg (type "llsf_msgs.Workpiece") (ptr ?p) (rcvd-at $?rcvd-at)
+											 (rcvd-from ?from-host ?from-port) (rcvd-via ?via))
+  =>
+  (retract ?mf) ; message will be destroyed after rule completes
+
+  (bind ?id (pb-field-value ?p "id"))
+	(bind ?at-machine-str (pb-field-value ?p "at_machine"))
+	(bind ?visible (if (pb-has-field ?p "visible") then (pb-field-value ?p "visible") else FALSE))
+
+	(printout t "Got " ?id " at " ?at-machine-str " visible? " ?visible crlf)
+
+  (assert (mps-status-feedback (sym-cat ?at-machine-str) BARCODE ?id))
+
+)
+
+(defrule net-recv-WorkpieceAddRing
+  ?mf <- (protobuf-msg (type "llsf_msgs.WorkpieceAddRing") (ptr ?p) (rcvd-at $?rcvd-at)
+											 (rcvd-from ?from-host ?from-port) (rcvd-via ?via))
+  =>
+  (retract ?mf) ; message will be destroyed after rule completes
+
+  (bind ?id (pb-field-value ?p "id"))
+	(bind ?ring-color (sym-cat (pb-field-value ?p "ring_color")))
+
+	(printout t "Add ring " ?ring-color " to workpiece " ?id crlf)
+
+	(do-for-fact ((?wp workpiece)) (eq ?wp:id ?id)
+		(printout t "Add ring " ?ring-color " to workpiece " ?id " *** " crlf)
+	  (modify ?wp (ring-colors (append$ ?wp:ring-colors ?ring-color)))
+	)
+)
 
+(deffunction net-create-WorkpieceInfo ()
+  (bind ?wi (pb-create "llsf_msgs.WorkpieceInfo"))
+
+  (do-for-all-facts
+    ((?wp workpiece)) TRUE
+
+    (bind ?w (pb-create "llsf_msgs.Workpiece"))
+		(pb-set-field ?w "id" ?wp:id)
+		(pb-set-field ?w "at_machine" (str-cat ?wp:at-machine))
+		(pb-set-field ?w "team" (str-cat ?wp:team))
+		(pb-set-field ?w "visible" ?wp:visible)
+
+		(if (neq ?wp:base-color nil) then
+			(pb-set-field ?w "base_color" (str-cat ?wp:base-color)))
+		(foreach ?rc ?wp:ring-colors (pb-add-list ?w "ring_colors" (str-cat ?rc)))
+		(if (neq ?wp:cap-color nil) then
+			(pb-set-field ?w "cap_color" (str-cat ?wp:cap-color)))
+
+		(pb-add-list ?wi "workpieces" ?w) ; destroys ?w
+  )	
+
+  (return ?wi)
+)
+
+(defrule net-send-WorkpieceInfo
+  (time $?now)
+  ?f <- (signal (type workpiece-info) (time $?t&:(timeout ?now ?t ?*WORKPIECEINFO-PERIOD*)) (seq ?seq))
+  (gamestate (cont-time ?ctime))
+  =>
+  (modify ?f (time ?now) (seq (+ ?seq 1)))
+  (bind ?wi (net-create-WorkpieceInfo))
+
+  (do-for-all-facts ((?client network-client)) (not ?client:is-slave)
+    (pb-send ?client:id ?wi))
+  (pb-destroy ?wi)
+)
```

Comment:
Good question, so that was a part of the implementation by @timn that I left in place. I think that is an aspect of tracking that we might still need to discuss the reasoning behind it. Options we have
1- Send to all info to all clients, allowing game theoretic reasoning and full knowledge of what the other team is doing
2- Send only team-specific workpiece info to each client on the private channel allowing them for partial reasoning about the success of their operations. 
3- Do not send any workpiece information and only use it internally for tracking

1 and 2 could potentially have the disadvantage of influencing the team's reasoning with inaccurate information if tracking failed for any reason (for example, a failed reader at an mps)

I think for now we should conceal the workpiece information (pick 3).
What do you think?

Reply:
I made it configurable!